issuance/issuance.go

```diff
@@ -366,37 +366,94 @@ func (p *Profile) generateTemplate(clk clock.Clock) *x509.Certificate {
 // that this certificate can be used for issuance. It also provides the .ID()
-// method, which returns an internal issuer ID for this certificate.
+// and .NameID() methods, which return internal issuer IDs for this certificate.
 type Certificate struct {
 	*x509.Certificate
+	id       IssuerID
+	nameID   IssuerNameID
+	nameHash [20]byte
+	keyHash  [20]byte
 }
 
-type IssuerID int64
+// NewCertificate wraps an in-memory cert in an issuance.Certificate, marking it
+// as an issuer cert. It also precomputes various hashes over the cert (exposed
+// by methods on this type), which may fail.
+func NewCertificate(ic *x509.Certificate) (*Certificate, error) {
+	res := Certificate{Certificate: ic}
+
+	// Compute ic.ID()
+	h := sha256.Sum256(ic.Raw)
+	res.id = IssuerID(big.NewInt(0).SetBytes(h[:4]).Int64())
+
+	// Compute ic.NameID()
+	res.nameID = truncatedHash(ic.RawSubject)
+
+	// Compute ic.NameHash()
+	res.nameHash = sha1.Sum(ic.RawSubject)
+
+	// Compute ic.KeyHash()
+	// The issuerKeyHash in OCSP requests is constructed over the DER encoding of
+	// the public key per RFC6960 (defined in RFC4055 for RSA and RFC5480 for
+	// ECDSA). We can't use MarshalPKIXPublicKey for this since it encodes keys
+	// using the SPKI structure itself, and we just want the contents of the
+	// subjectPublicKey for the hash, so we need to extract it ourselves.
+	var spki struct {
+		Algorithm pkix.AlgorithmIdentifier
+		PublicKey asn1.BitString
+	}
+	_, err := asn1.Unmarshal(ic.RawSubjectPublicKeyInfo, &spki)
+	if err != nil {
+		return nil, err
+	}
+	res.keyHash = sha1.Sum(spki.PublicKey.RightAlign())
+
+	return &res, nil
+}
 
 // ID provides a stable ID for an issuer's certificate. This is used for
 // identifying which issuer issued a certificate in the certificateStatus table.
 // This value is computed as a truncated hash over the whole certificate,
 // meaning it is highly unique but not computable from end-entity certs.
 func (ic *Certificate) ID() IssuerID {
-	h := sha256.Sum256(ic.Raw)
-	return IssuerID(big.NewInt(0).SetBytes(h[:4]).Int64())
+	return ic.id
 }
 
-// IssuerNameID is a statistically-unique small ID which can be computed from
-// both CA and end-entity certs to link them together into a validation chain.
-// It is computed as a truncated hash over the issuer Subject Name bytes, or
-// over the end-entity's Issuer Name bytes, which are required to be equal.
-// TODO(#5152): Rename this "IssuerID" when we've fully deprecated the old-style
-// IssuerIDs and replaced them with NameIDs.
-type IssuerNameID int64
-
 // NameID computes the IssuerNameID from an issuer certificate, i.e. it
 // computes a truncated hash over the issuer's Subject Name raw bytes. Useful
 // for storing as a lookup key in contexts that don't expect hash collisions.
 func (ic *Certificate) NameID() IssuerNameID {
-	return truncatedHash(ic.RawSubject)
+	return ic.nameID
+}
+
+// NameHash computes the SHA1 hash over the issuer certificate's Subject
```

Comment:
Is it accurate to say that these three functions compute the hash since they are precomputed? Maybe to the caller of the function it is irrelevant.

Reply:
Yeah, I left them that way because whether they're computed at call time or precomputed should be an internal implementation detail opaque to the caller. I suppose they could just say "returns", though.

Reply:
Actually yeah I've updated a few of the comments to be clearer and more correct. Take a look?

Reply:
Awesome, approved!